### ISSUE TITLE
^compile does not resolve internal `dependsOn` project dependency

### ISSUE THREAD

**Issue by jeffreyolchovy** [sbt/sbt#3392]
## steps
Given a project definition which includes both an sbt plugin and a non-plugin project that the sbt plugin depends on:

```scala
inThisBuild(
  Seq(
    organization := "com.example",
    version := "0.1"
  )
)

lazy val root = (project in file("."))
  .aggregate(library, plugin)
  .enablePlugins(CrossPerProjectPlugin)

val library = (project in file("library"))
  .settings(
    scalaVersion := "2.10.6",
    crossScalaVersions := Seq("2.10.6", "2.12.3")
  )

val plugin = (project in file("plugin"))
  .settings(
    sbtPlugin := true,
    crossSbtVersions := Seq("0.13.16", "1.0.0-RC3")
  )
  .dependsOn(library)
```

## problem
When attempting to cross-compile the plugin for the given sbt versions, the following error occurs during dependency resolution:

```
> ^compile
[info] Setting `sbtVersion in pluginCrossBuild` to 0.13.16
[info] Set current project to plugin (in build file:/Users/jeffo/Projects/jvm/sbt-issue/)
[success] Total time: 0 s, completed Jul 31, 2017 9:58:42 AM
[info] Setting `sbtVersion in pluginCrossBuild` to 1.0.0-RC3
[info] Set current project to plugin (in build file:/Users/jeffo/Projects/jvm/sbt-issue/)
[info] Resolving com.example#library_2.12;0.1 ...
[warn] 	module not found: com.example#library_2.12;0.1
[logging elided] ...
[info] Resolving jline#jline;2.14.3 ...
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	::          UNRESOLVED DEPENDENCIES         ::
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	:: com.example#library_2.12;0.1: not found
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn] 	Note: Unresolved dependencies path:
[warn] 		com.example:library_2.12:0.1
[warn] 		  +- com.example:plugin:0.1 (scalaVersion=2.12, sbtVersion=1.0)
[trace] Stack trace suppressed: run last plugin/*:update for the full output.
[error] (plugin/*:update) sbt.ResolveException: unresolved dependency: com.example#library_2.12;0.1: not found
```

## expectation
I would expect that when issuing `^compile`, the internal projects that my plugin depends on would be available to it without have to do a `publishLocal` or `+publishLocal`.

## notes
The `^compile` task succeeds if I manually issue `+publishLocal` from the `library` project before attempting `^compile`.

sbt version: 0.13.16, cross building for 0.13.16 and 1.0.0-RC3

**Comment by dwijnand** [comment]
Strange. This should've been solved in #3356.

**Comment by dwijnand** [comment]
Oh! Cos #3356 only changes `^^` (aka pluginSwitch) and not `^`/`pluginCross`.

**Comment by dwijnand** [comment]
No it's not that. Even the setting of scalaVersion in `^^` is too shallow: it only sets the scalaVersion in the Global scope and the current project scope, which is insufficient for the setup in your repro.

I refer to my notes [here](https://github.com/sbt/sbt/issues/3205#issuecomment-311356035) about how existing functionality (`++` and `set every`) works.

**Comment by jeffreyolchovy** [comment]
So is the current functionality the desired functionality for sbt?

**Comment by dwijnand** [comment]
Not in my book - given that there's a possible manual translation to make that legitimate build work for sbt 1, `^` and `^^` should work as you expected it to.

If your question is because you're stuck, then there are ways to avoid the problem. I haven't hit this particular bug because I set my scalaVersion in terms of sbtVersion in pluginCrossBuild: https://github.com/dwijnand/sbt-project-graph/blob/v0.2.1/build.sbt#L16-L20

**Comment by jeffreyolchovy** [comment]
Nope, not stuck, just curious. Used the same workaround, but I think I sourced it from something else of yours (maybe another open issue or a different plugin).

**Comment by eed3si9n** [comment]
To avoid adding `scalaVersion` at the project level by default, in #3356 I started appending `scalaVersion` at the project level only for _the current subproject_. In the hindsight I was walking into the situation like this whenever there's a plugin that depends on other subprojects.

We could try issuing `++2.x.y` inside `^^` command as fix.

**Comment by jeffreyolchovy** [comment]
@eed3si9n I had tried that, but unfortunately I ran into another issue which forced me to reinstate the `scalaVersion` on the subprojects. I am still trying to track down exactly what is the cause of this other issue, but it's something like:

```
- projectA is a plugin
- projectB is a plugin
- projectA and projectB both cross build for 0.13 and 1.0
- projectA and projectB do not set scalaVersion
- projectA and projectB both use scripted
- projectB depends on projectA using `dependsOn`
```

When running `^compile` for projectB, I would get:
```
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	::          UNRESOLVED DEPENDENCIES         ::
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	:: org.scala-sbt#scripted-sbt_2.10;1.0.0-RC3: not found
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn] 	Note: Unresolved dependencies path:
[warn] 		org.scala-sbt:scripted-sbt_2.10:1.0.0-RC3 ((sbt.ScriptedPlugin) ScriptedPlugin.scala#L104)
[warn] 		  +- com.example:project-a:0.1.0 (scalaVersion=2.10, sbtVersion=1.0)
```

The way I worked around this was not using `dependsOn`, but just adding a `libraryDependency` (and issuing `publishLocal`) to projectA on projectB like I did for the library and plugin `dependsOn` issue above.

I believe this only happens when I have the `scripted` dependency, so maybe I should be doing something different in my `plugins.sbt` when attempting to cross-build for 0.13 and 0.1. Currently, my `plugins.sbt` has the standard:

```
libraryDependencies += "org.scala-sbt" % "scripted-plugin" % sbtVersion.value
```

UPDATE:
Seems like this other issue would be fixed by https://github.com/sbt/sbt/issues/3325

**Comment by qayshp** [comment]
I'm trying to get a simpler repro, but I seem to be hitting either this or something that looks similar with a Scala JS build.
Any dependsOn that I add also show up as unresolved until I do a publishLocal on the specific sub-project.
